### PR TITLE
Deserialize series using field names, not order

### PR DESCRIPTION
**Pull request by SafariMonkey**
## Description

Add a custom `Deserialize` impl for `Series` which uses the field names in the columns field of the response to allow the inner type to deserialize itself as a map instead of relying on the field order. This means that the field order of your structs do not need to match the query field order, and prevents mistakes due to inconsistencies between those. It also means a `SELECT *` could go into a hashmap and work just fine if you wanted to.

I added a couple of crates while I was prototyping: `smol_str` and `maplit` (dev). These can easily be removed, but I found a 10% speedup on deserialising+serving (mocking the databse response) when using all `SmolStr` over `String` in both headers and my field values, and `hashmap` literals are nice in tests. (It's probably worth noting that the SmolStr change improved performance from 1700 to 1900 requests per second, but actually hitting the database it was more like a difference of 213 to 218 - certainly nothing to write home about.)

If you were interested in improving the performance, though, I imagine that the Results struct _could_ hold the full response body + a `Vec` of `serde_json::value::RawValue`, and then you could also have a `deserialize_next_borrowed<'de, T>(&'de self) -> Result<Return<T>, Error> where T: Deserialize<'de> + Send` or something. I experimented with such an implementation on [the `deserialize-borrowed` branch](https://github.com/SafariMonkey/influxdb-rust/compare/deserialize-using-column-names...SafariMonkey:deserialize-borrowed). I didn't go the whole way with the RawValues borrowing from the body, but using something like `owning_ref` it should be possible - it requires unsafe with the current API of `owning_ref`, though, and speaking briefly with the creator of that project today, they made it clear that they are not maintaining it at this time. I don't know if an API that copies each entire query result is really a win vs. an API that deserialises it into a structured JSON representation instead, especially as the current (structured JSON) one can just reuse the string allocations from the JSON in the result. Using the borrowing based API is also more awkward, of course, as the `DatabaseQueryResult` can't be a temporary... it's probably a case of overengineering.

In summary, I think the core of this PR is I think a correctness win without an API change. There are some performance improvements possible, but given the slowness of InfluxDB, they're unlikely to make enough of a difference to be noticeable.

### Checklist
- [x] Formatted code using `cargo fmt --all`
- [x] Linted code using clippy `cargo clippy --all-targets --all-features -- -D warnings` (first commit in PR)
- [ ] Updated README.md using `cargo readme > README.md` (no, because it mostly just deleted stuff)
- [x] Reviewed the diff. Did you leave any print statements or unnecessary comments?
- [x] Any unfinished work that warrants a separate issue captured in an issue with a TODO code comment (none)

### CODE REVIEW

**Comment by SafariMonkey** [comment]
I made a [further branch](https://github.com/SafariMonkey/influxdb-rust/compare/deserialize-using-column-names...SafariMonkey:deserialize-tagged) after realising that the current design does not handle `GROUP BY` correctly. For now I chose to copy the entire `Return` and `Series` structs as well as the entrypoint to the `Deserialize` implementation. This feels like a nicer API to work with than one that's generic over tagged-ness, at the expense of some copying in the code. The API can of course be changed, and struct/method names in particular are subject to bikeshedding.

My API experiment sped up by around 70% by using a "group by" on the three tags I have, presumably due to the reduction in redundancy in the payload.

Oh, and I just ran into an issue: if you don't get any results, `"series"` isn't returned. [This fixes that.](https://github.com/SafariMonkey/influxdb-rust/compare/deserialize-tagged...SafariMonkey:series-default)

**Comment by Empty2k12** [comment]
I have played around with the changes and am quite happy with them! I will merge when you give the final go, that you're happy with everything at its current status.

> You are the author of this crate, and whatever approach is taken I want it to be one you are happy with.

I find this project to be more of a community effort to create a modern and easy to use InfluxDb client, than my personal thing. I'm happy to see contributions and progress, everything else is decided together!

**Comment by cryptoquick** [comment]
I just checked this one out, and it seems it double-quotes strings, at least in my case, just FYI. I did run into the problem it solves, and, otherwise, it seems to be a pretty drop-in replacement.

**Comment by cryptoquick** [comment]
Hi, is there any way this branch can be merged into master? It does work, and since there's a fix in the other branch for double quotes, hopefully this solves both issues, yeah?

@SafariMonkey Can you merge master into this branch and give us the final go? If you like, I can give you a thumbs up from my codebase.

**Comment by SafariMonkey** [comment]
Hi, sorry for the long silence.

Life things have been getting in the way of my being able to go in and figure out how to address the issue raised by @cryptoquick. I keep meaning to pick it up but not getting to it - I was hoping to finally do so now that I'm on holiday.

I appreciate your patience, as well as the encouragement to get it done :) . I did not realize that the quoting issue had already been independently resolved, in which case it's basically good to go - that was my only outstanding concern, so I'll just rebase it, confirm it works with my code, and then give it a once over before approving.

By the way, the way this library does data insertion doesn't work with my use case at all, as we have custom quoting requirements. (We have our own escaping system to deal with InfluxDB's ridiculous quoting issues.) Fortunately, I'm not looking to use this library to insert data, so as long as it gives me the raw form of the metrics and doesn't try to escape them on the way out itself, it should work fine.

**Comment by SafariMonkey** [comment]
I've rebased, tried it out, and it seems to work. I've read through the changes again and they look fine, so if you are happy with them too, feel free to merge them.

I'll open a PR for the `GROUP BY` support too.

Thank you for your patience.

**Comment by Empty2k12** [comment]
Thank you again for this contribution! Will merge now!

**Comment by SafariMonkey** [comment]
As promised, I posted the `GROUP BY` support PR: #69.

**Comment by cryptoquick** [comment]
BTW, we're using this code in our production branch it works perfectly. Great work! ❤️
